std/stdio.d

```diff
@@ -2920,7 +2919,7 @@ $(D Range) that locks the file and allows fast writing to it.
                     {
                         char[4] buf;
                         immutable size = encode!(UseReplacementDchar.yes)(buf, c);
-                        foreach (i ; 0 .. size)
+                        for (int i; i < size; ++i)
```

Comment:
Is `for` really faster than `foreach` here?

Reply:
foreach makes three extra `int`s

```d
    int size = 5;
    foreach (i; 0 .. size)
    {
    }
    
    for (int j = 0; j < size; ++j)
    {
    }
```

becomes

```
int size = 5;
{
	int __key46 = 0;
	int __limit47 = size;
	for (; __key46 < __limit47; __key46 += 1)
	{
		int i = __key46;
	}
}
{
	int j = 0;
	for (; j < size; j += 1)
	{
		{
		}
	}
}
```

Reply:
Does that actually affect generated code, though?

Reply:
My knowledge of ASM is limited, but it sure looks like it does

`foreach`
```asm
.text._Dmain    segment
    assume  CS:.text._Dmain
_Dmain:
        push    RBP
        mov RBP,RSP
        sub RSP,010h
        mov EAX,5
        mov -010h[RBP],EAX
        mov dword ptr -0Ch[RBP],0
        mov -8[RBP],EAX
L1A:        mov ECX,-0Ch[RBP]
        cmp ECX,-8[RBP]
        jge L2D
        mov EDX,-0Ch[RBP]
        mov -4[RBP],EDX
        inc dword ptr -0Ch[RBP]
        jmp short   L1A
L2D:        xor EAX,EAX
        leave
        ret
        add [RAX],AL
        add [RAX],AL
.text._Dmain    ends
```

`for`
```asm
.text._Dmain    segment
    assume  CS:.text._Dmain
_Dmain:
        push    RBP
        mov RBP,RSP
        sub RSP,010h
        mov dword ptr -8[RBP],5
        mov dword ptr -4[RBP],0
L16:        mov EAX,-4[RBP]
        cmp EAX,-8[RBP]
        jge L23
        inc dword ptr -4[RBP]
        jmp short   L16
L23:        xor EAX,EAX
        leave
        ret
        add 0[RCX],AL
.text._Dmain    ends
```

Reply:
Yeah, the number of `int` declarations may not have any correlation with the actual number of stack variables allocated by the asm. Given the simplicity of the loops above, I'm expecting that even dmd would enregister the loop variables, so the argument about the number of `int` declarations isn't even relevant.

Reply:
@quickfur , Jack posted some asm above, do you not see his comment? https://github.com/dlang/phobos/pull/6091#discussion_r164501971

@JackStouffer I'm guessing that`s without `-O`? Phobos is built with `-O -release`, so we should be comparing disassemblies from code built with those flags.

Reply:
>Again, please show the asm to prove that the for-loop is better than the foreach loop.

I'm unclear as to what ASM you're asking for; I thought I just posted it.

Reply:
Sorry, I must have been looking at a stale page when I wrote my comment.

But yeah, it does look like that's without `-O`. For performance comparisons `dmd -O` is really the bare minimum; I would actually prefer to see what gdc/ldc do with the same code, because they have better loop optimizations than dmd. (Though in the latter case the challenge becomes how to prevent overly-aggressive optimizations like in ldc, where the optimizer just runs the whole thing at compile-time and substitutes the answer directly. :-P)